opm/material/densead/FastSmallVector.hpp

```diff
@@ -0,0 +1,164 @@
+#include <array>
+#include <cstring>
+
+namespace Opm {
+namespace DenseAd {
+
+template <typename ValueType, unsigned N>
+class FastSmallVector
+{
+public:
+
+    FastSmallVector();
+    explicit FastSmallVector(const size_t num_elem);
+    FastSmallVector(const size_t num_elem, const ValueType value);
+    FastSmallVector(const FastSmallVector& other);
+    FastSmallVector(FastSmallVector&& other);
+    ~FastSmallVector();
+
+    FastSmallVector& operator=(const FastSmallVector& other);
+    FastSmallVector& operator=(FastSmallVector&& other);
+
+    ValueType& operator[](size_t index);
+    const ValueType& operator[](size_t index) const;
+    size_t size() const;
+
+private:
+    std::array<ValueType, N> small_buffer_;
+    std::size_t size_;
+    ValueType* data_ptr_;
+};
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector()
+    : size_(0)
+    , data_ptr_(small_buffer_.data())
+{
+}
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector(const size_t num_elem)
+    : size_(num_elem)
+    , data_ptr_(small_buffer_.data())
+{
+    if (size_ > N) {
+        data_ptr_ = new ValueType[num_elem];
+    }
+}
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector(const size_t num_elem, const ValueType value)
+    : FastSmallVector(num_elem)
+{
+    std::fill_n(data_ptr_, size_, value);
+}
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector(FastSmallVector&& other)
+   : size_ (other.size_)
+{
+    small_buffer_ = std::move(other.small_buffer_);
+    if (size_ <= N) {
+        data_ptr_ = small_buffer_.data();
+    } else {
+        data_ptr_ = other.data_ptr_;
+    }
+
+    other.data_ptr_= nullptr;
+    other.size_ = 0;
+}
+
+template <typename ValueType, unsigned N>
+FastSmallVector<ValueType, N>::
+FastSmallVector(const FastSmallVector& other)
+   : small_buffer_(other.small_buffer_)
+   , size_(other.size_)
+{
+    if (size_ <= N) {
+        data_ptr_ = small_buffer_.data();
+    } else {
+        data_ptr_ = new ValueType[size_];
```

Comment:
you could save a bit of memory if you only use `data_ptr_` for the entries that do not fit into `small_buffer_`. the performance implications of this should be measured, though.

Reply:
Then we need to check whether the data we are trying to access is in data_ptr_ or in the small_buffer_. Both ways have their pros and cons. I suggest we stay with the current way.

Reply:
okay. These kinds of objects are not intended to be stored in large arrays anyway? (i.e., using them for e.g. the intensive quantities would probably make the memory requirements explode...)